// @quintype/copyright-header/dist/src/config/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LINE_READER_LIMIT = exports.SUPPORTED_LANGUAGES = exports.COPYRIGHT_KEYWORDS = void 0;
const COPYRIGHT_KEYWORDS = ['copyright', '(c)', '©'];
exports.COPYRIGHT_KEYWORDS = COPYRIGHT_KEYWORDS;
const LINE_READER_LIMIT = 50;
exports.LINE_READER_LIMIT = LINE_READER_LIMIT;
const SUPPORTED_LANGUAGES = [
    {
        name: 'C',
        fileExtensions: ['c'],
        commentSyntax: {
            singleLine: {
                start: ['//']
            },
            multiLine: {
                start: ['/*'],
                end: ['*/']
            }
        }
    },
    {
        name: 'CPP',
        fileExtensions: ['cpp'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Javascript',
        fileExtensions: ['js'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Javascript React',
        fileExtensions: ['jsx'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Typescript',
        fileExtensions: ['ts'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Typescript React',
        fileExtensions: ['tsx'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Go',
        fileExtensions: ['go'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Java',
        fileExtensions: ['java'],
        commentSyntax: {
            extendFromLanguage: 'C'
        }
    },
    {
        name: 'Clojure',
        fileExtensions: ['clj'],
        commentSyntax: {
            singleLine: {
                start: [';']
            }
        }
    },
    {
        name: 'Python',
        fileExtensions: ['py'],
        commentSyntax: {
            singleLine: {
                start: ['#']
            }
        }
    }
    // Add support for new programming language here
];
exports.SUPPORTED_LANGUAGES = SUPPORTED_LANGUAGES;
